pilot/pkg/config/kube/ior/route.go

```diff
@@ -627,3 +367,155 @@ func hostHash(name string) string {
 	hash := sha256.Sum256([]byte(name))
 	return hex.EncodeToString(hash[:8])
 }
+
+func (r *route) reconcileGateway(config *config.Config, routes *v1.RouteList) error {
+	gateway, ok := config.Spec.(*networking.Gateway)
+
+	if !ok {
+		return fmt.Errorf("could not decode spec as Gateway from %v", config)
+	}
+
+	var serviceNamespace string
+	var serviceName string
+	var err error
+
+	serviceNamespace, serviceName, err = r.findService(gateway)
+
+	if err != nil {
+		return errors.Wrapf(err, "gateway %s/%s does not specify a valid service", config.Namespace, config.Name)
+	}
+
+	routeMap := make(map[string]v1.Route)
+
+	for _, v := range routes.Items {
+		routeMap[v.Name] = v
+	}
+
+	var result *multierror.Error
+
+	for _, server := range gateway.Servers {
+		for _, host := range server.Hosts {
+			var route *v1.Route
+			var err error
+
+			name := getRouteName(config.Namespace, config.Name, host)
+
+			_, found := routeMap[name]
+
+			if found {
+				route, err = r.updateRoute(config.Meta, host, server.Tls, serviceNamespace, serviceName)
+			} else {
+				route, err = r.createRoute(config.Meta, host, server.Tls, serviceNamespace, serviceName)
+			}
+
+			if err != nil {
+				result = multierror.Append(result, err)
+			} else {
+				delete(routeMap, route.Name)
+			}
+		}
+	}
+
+	for k, v := range routeMap {
+		IORLog.Debugf("clean up route %s for host %s", k, getHost(v))
+		if err := r.deleteRoute(&v); err != nil {
+			result = multierror.Append(result, err)
+		}
+	}
+
+	return result.ErrorOrNil()
+}
+
+func (r *route) processEvent(old, curr *config.Config, event model.Event) error {
+	if IORLog.GetOutputLevel() >= log.DebugLevel {
+		debugMessage := fmt.Sprintf("event %v arrived:", event)
+		if event == model.EventUpdate {
+			debugMessage += fmt.Sprintf("\told object: %v", old)
+		}
+		debugMessage += fmt.Sprintf("\tnew object: %v", curr)
+
+		IORLog.Debug(debugMessage)
+	}
+
+	var (
+		err       error
+		isManaged bool
+	)
+
+	isManaged, err = isManagedByIOR(*curr)
+
+	if err != nil {
+		return err
+	}
+
+	if !isManaged {
+		IORLog.Debugf("skipped processing routes for gateway %s/%s, as it is annotated by user", curr.Name, curr.Namespace)
+		return nil
+	}
+
+	config := r.store.Get(collections.IstioNetworkingV1Alpha3Gateways.Resource().GroupVersionKind(), curr.Name, curr.Namespace)
+
+	// Stop early
+	if config != nil && curr.ResourceVersion < config.ResourceVersion {
```

Comment:
ResourceVersion is "opaque", meaning that you shouldn't depend on the (current) format of its contents. You're only allowed to check if two resourceVersions are equal or not (in other words: you can do `resourceVersion1 == resourceVersion2` and nothing else).

Yes, currently, this is an integer, but this is an implementation detail of etcd AFAIK. It might change in the future. Also, since resourceVersion is a string, comparing them with `<` doesn't even work correctly (for example: `"12" < "2"`).

Reply:
Actually, I think you shouldn't fetch the latest config at all. Just use `curr`.

Reply:
Still need to fetch current to determinate if the `Gateway` still exists. For reconciliation, `curr` is used.